From 49a5241ce874c374fe736417ee4e4a423f402e0d Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:27:06 -0400
Subject: [tools] perf/traceevent: Update printk formats when entered

Message-id: <1408462094-14194-258-git-send-email-jolsa@redhat.com>
Patchwork-id: 88260
O-Subject: [PATCH RHEL7.1 BZ1131394 257/325] tools lib traceevent: Update printk formats when entered
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 18900af8292180151c82f0762506fa0740aa54a5
Author: Steven Rostedt (Red Hat) <rostedt@goodmis.org>
Date: Fri Nov 1 17:53:54 2013 -0400

description
===========
Instead of cropping off the '"' and '\n"' from a printk format every
time it is referenced, do it when it's added. This makes it easier to
reference a printk_map and should speed things up a little.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/lib/traceevent/event-parse.c b/tools/lib/traceevent/event-parse.c
index deedff9..856b791 100644
--- a/tools/lib/traceevent/event-parse.c
+++ b/tools/lib/traceevent/event-parse.c
@@ -604,10 +604,11 @@ find_printk(struct pevent *pevent, unsigned long long addr)
  * This registers a string by the address it was stored in the kernel.
  * The @fmt passed in is duplicated.
  */
-int pevent_register_print_string(struct pevent *pevent, char *fmt,
+int pevent_register_print_string(struct pevent *pevent, const char *fmt,
      unsigned long long addr)
 {
  struct printk_list *item = malloc(sizeof(*item));
+ char *p;
 
  if (!item)
   return -1;
@@ -615,10 +616,21 @@ int pevent_register_print_string(struct pevent *pevent, char *fmt,
  item->next = pevent->printklist;
  item->addr = addr;
 
+ /* Strip off quotes and '\n' from the end */
+ if (fmt[0] == '"')
+  fmt++;
  item->printk = strdup(fmt);
  if (!item->printk)
   goto out_free;
 
+ p = item->printk + strlen(item->printk) - 1;
+ if (*p == '"')
+  *p = 0;
+
+ p -= 2;
+ if (strcmp(p, "\\n") == 0)
+  *p = 0;
+
  pevent->printklist = item;
  pevent->printk_count++;
 
@@ -3887,7 +3899,6 @@ get_bprint_format(void *data, int size __maybe_unused,
  struct format_field *field;
  struct printk_map *printk;
  char *format;
- char *p;
 
  field = pevent->bprint_fmt_field;
 
@@ -3909,20 +3920,8 @@ get_bprint_format(void *data, int size __maybe_unused,
   return format;
  }
 
- p = printk->printk;
- /* Remove any quotes. */
- if (*p == '"')
-  p++;
- if (asprintf(&format, "%s : %s", "%pf", p) < 0)
+ if (asprintf(&format, "%s : %s", "%pf", printk->printk) < 0)
   return NULL;
- /* remove ending quotes and new line since we will add one too */
- p = format + strlen(format) - 1;
- if (*p == '"')
-  *p = 0;
-
- p -= 2;
- if (strcmp(p, "\\n") == 0)
-  *p = 0;
 
  return format;
 }
diff --git a/tools/lib/traceevent/event-parse.h b/tools/lib/traceevent/event-parse.h
index 7503edf..9ab6367 100644
--- a/tools/lib/traceevent/event-parse.h
+++ b/tools/lib/traceevent/event-parse.h
@@ -533,7 +533,7 @@ int pevent_register_comm(struct pevent *pevent, const char *comm, int pid);
 void pevent_register_trace_clock(struct pevent *pevent, char *trace_clock);
 int pevent_register_function(struct pevent *pevent, char *name,
         unsigned long long addr, char *mod);
-int pevent_register_print_string(struct pevent *pevent, char *fmt,
+int pevent_register_print_string(struct pevent *pevent, const char *fmt,
      unsigned long long addr);
 int pevent_pid_is_registered(struct pevent *pevent, int pid);
 
-- 
1.7.1